src/axom/spin/BVH.hpp

```diff
@@ -0,0 +1,848 @@
+   *
+   * \pre min != nullptr
+   * \pre max != nullptr
+   */
+  void getBounds( FloatType* min, FloatType* max ) const;
+
+  /*!
+   * \brief Finds the candidate geometric entities that contain each of the
+   *  given query points.
+   *
+   * \param [out] offsets offset to the candidates array for each query point
+   * \param [out] counts stores the number of candidates per query point
+   * \param [out] candidates array of the candidate IDs for each query point
+   * \param [in]  numPts the total number of query points supplied
+   * \param [in]  x array of x-coordinates
+   * \param [in]  y array of y-coordinates
+   * \param [in]  z array of z-coordinates, may be nullptr if 2D
+   *
+   * \note offsets and counts are pointers to arrays of size numPts that are
+   *  pre-allocated by the caller before calling find().
+   *
+   * \note The candidates array is allocated internally by the method and
+   *  ownership of the memory is transferred to the caller. Consequently, the
+   *  caller is responsible for properly deallocating the candidates buffer.
+   *
+   * \note Upon completion, the ith query point has:
+   *  * counts[ i ] candidates
+   *  * Stored in the candidates array in the following range:
+   *    [ offsets[ i ], offsets[ i ]+counts[ i ] ]
+   *
+   * \pre offsets != nullptr
+   * \pre counts  != nullptr
+   * \pre candidates == nullptr
+   * \pre x != nullptr
+   * \pre y != nullptr if dimension==2 || dimension==3
+   * \pre z != nullptr if dimension==3
+   */
+  /// @{
+  void find( IndexType* offsets, IndexType* counts, IndexType*& candidates,
+             IndexType numPts, const FloatType* x, const FloatType* y ) const;
+  void find( IndexType* offsets, IndexType* counts,
+             IndexType*& candidates, IndexType numPts,
+             const FloatType* x, const FloatType* y, const FloatType* z ) const;
+  /// @}
+
+  /*!
+   * \brief Writes the BVH to the specified VTK file for visualization.
+   * \param [in] fileName the name of VTK file.
+   * \note Primarily used for debugging.
+   */
+  void writeVtkFile( const std::string& fileName ) const;
+
+private:
+
+/// \name Private Members
+/// @{
+
+  IndexType m_numItems;
+  const FloatType* m_boxes;
+  internal::linear_bvh::BVHData< FloatType,NDIMS > m_bvh;
+
+/// @}
+
+  DISABLE_COPY_AND_ASSIGNMENT(BVH);
+  DISABLE_MOVE_AND_ASSIGNMENT(BVH);
+};
+
+//------------------------------------------------------------------------------
+//  BVH IMPLEMENTATION HELPER METHODS
+//------------------------------------------------------------------------------
+namespace
+{
+
+//------------------------------------------------------------------------------
+template< typename FloatType >
+AXOM_HOST_DEVICE
+bool InLeft( const internal::linear_bvh::Vec< FloatType, 3>& point,
+             const internal::linear_bvh::Vec< FloatType, 4>& s1,
+             const internal::linear_bvh::Vec< FloatType, 4>& s2 )
+{
+  bool in_left = true;
+  if ( point[0]  < s1[0] ) in_left = false;
```

Comment:
Nitpick but I think it's cleaner (and potentially faster) to have
``` c++
if ( cond ) return false;
...
return true;
```

Reply:
Probably faster on the CPU (but maybe not e.g. no branch prediction) to return early but slower on the GPU. Pick your poison.

https://godbolt.org/z/1FrHN9